### PR TITLE
GH-39013: [Go][Integration] Support cABI import/export of StringView

### LINKED ISSUE #39013: [Go][Integration] Add support for C abi Utf8View to the go implementation
### Describe the enhancement requested

After https://github.com/apache/arrow/pull/35769 , we can add support for importing/exporting arrays through the C abi and test between C++ and go

### Component(s)

Go, Integration

### PR DESCRIPTION
**Pull request by bkietz**
### Rationale for this change

The Go implementation should support import/export of the new data types. This will enable integration testing between the C++ and Go implementations.

### What changes are included in this PR?

Added import/export for the new data types and arrays of data of those types.

### Are these changes tested?

Yes, they will be covered by the integration tests and existing Go unit tests.

### Are there any user-facing changes?

This is a user facing change

* Closes: #39013

### CODE REVIEW

File: dev/archery/archery/integration/datagen.py [hunk 1]
```diff
@@ -927,6 +927,85 @@ class LargeListColumn(_BaseListColumn, _LargeOffsetsMixin):
...
+        super().__init__(name, nullable=nullable,
+                         metadata=metadata)
+        self.value_field = value_field
+
+    @property
+    def column_class(self):
+        return ListViewColumn
+
+    def _get_type(self):
+        return OrderedDict([
```
Comment (pitrou) [inline dev/archery/archery/integration/datagen.py:943]: FTR, we don't need `OrderedDict` anymore as regular dicts are now ordered by default.

File: dev/archery/archery/integration/runner.py [hunk 2]
```diff
@@ -193,6 +193,8 @@ def _run_test_cases(self,
         ``case_runner`` ran against ``test_cases``
         """
         def case_wrapper(test_case):
+            if serial:
+                return case_runner(test_case)
```
Comment (pitrou) [inline dev/archery/archery/integration/runner.py:197]: Is this addition deliberate?

**Review by pitrou — CHANGES_REQUESTED** [review]

File: dev/archery/archery/integration/runner.py (hunk 2 above)
Comment (bkietz) [reply dev/archery/archery/integration/runner.py:197]: Yes: printer.cork() was failing to dump output when the test segfaulted. In any case, corking the printer adds no value when we're running in serial.

File: dev/archery/archery/integration/datagen.py [hunk 3]
```diff
@@ -927,6 +927,85 @@ class LargeListColumn(_BaseListColumn, _LargeOffsetsMixin):
...
+    def generate_column(self, size, name=None):
+        MAX_LIST_SIZE = 4
+
+        is_valid = self._make_is_valid(size)
+        offsets = []
+        sizes = np.random.randint(0, MAX_LIST_SIZE + 1, size=size)
+        offset = 0
+        for s in sizes:
+            offsets.append(offset)
+            offset += int(s)
```
Comment (felipecrv) [inline dev/archery/archery/integration/datagen.py:959]: Offsets should probably be out of order to break code that assumes list-view offsets are sorted like in lists.

File: go/arrow/array/list.go [hunk 4]
```diff
@@ -1411,217 +1410,56 @@ func (b *baseListViewBuilder) UnmarshalJSON(data []byte) error {
...
-// Find the maximum offset+size in a LIST_VIEW array.
+// Find the minimum offset+size in a LIST_VIEW/LARGE_LIST_VIEW array.
 //
 // Pre-conditions:
 //
-//	input.DataType() is ListViewType
-//	input.Len() > 0 && input.NullN() != input.Len()
-func maxListViewOffset32(input arrow.ArrayData) int {
+//	input.DataType() is ListViewType if Offset=int32 or LargeListViewType if Offset=int64
+//	input.Len() > 0
```
Comment (felipecrv) [inline go/arrow/array/list.go:1418]: Why remove the `NullN() != Len()` pre-condition?

File: go/arrow/array/list.go [hunk 5]
```diff
@@ -1411,217 +1410,56 @@ func (b *baseListViewBuilder) UnmarshalJSON(data []byte) error {
...
-				}
-			}
+	for i -= 1; i >= 0; i -= 1 {
+		if maxEnd == maxLegalOffset {
+			// Fast path: the maximum offset+size is frequently exactly the end of the child array,
+			// and frequently a view which has this offset+size will be near the end of the array.
+			return maxEnd
+		}
+		if end := offsets[i] + sizes[i]; end > maxEnd {
+			maxEnd = end
```
Comment (felipecrv) [inline go/arrow/array/list.go:1462]: You can take a look at my final version of this function in the C++ implementation: 

https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/list_util.cc#L91

**Review by felipecrv — COMMENTED** [review]
I'm adding more comments.

File: go/arrow/cdata/cdata.go [hunk 6]
```diff
@@ -683,6 +743,43 @@ func (imp *cimporter) importListLike() (err error) {
...
+
+	var nulls, offsets, sizes *memory.Buffer
+	if nulls, err = imp.importNullBitmap(0); err != nil {
+		return
+	}
+	if nulls != nil {
+		defer nulls.Release()
+	}
+
+	if offsets, err = imp.importOffsetsBuffer(1, offsetSize); err != nil {
```
Comment (felipecrv) [inline go/arrow/cdata/cdata.go:765]: Is `importOffsetsBuffer` importing only `len` offsets instead of `len + 1`?

File: go/arrow/cdata/cdata_exports.go [hunk 7]
```diff
@@ -328,6 +336,15 @@ func allocateBufferPtrArr(n int) (out []*C.void) {
 	return
 }
 
+func allocateBufferSizeArr(n int) (out []C.int64_t) {
+	s := (*reflect.SliceHeader)(unsafe.Pointer(&out))
+	s.Data = uintptr(C.calloc(C.size_t(n), C.size_t(unsafe.Sizeof(int64(0)))))
+	s.Len = n
+	s.Cap = n
+
+	return
```
Comment (felipecrv) [inline go/arrow/cdata/cdata_exports.go:345]: is this necessary?

File: go/arrow/compute/fieldref.go [hunk 8]
```diff
@@ -346,17 +346,18 @@ func FieldRefList(elems ...interface{}) FieldRef {
 // NewFieldRefFromDotPath parses a dot path into a field ref.
 //
 // dot_path = '.' name
-//			| '[' digit+ ']'
-//			| dot_path+
+//
+//	| '[' digit+ ']'
+//	| dot_path+
```
Comment (felipecrv) [inline go/arrow/compute/fieldref.go:351]: the `|` should align with the `=` on the previous line

File: go/arrow/compute/fieldref.go [hunk 9]
```diff
@@ -282,31 +282,31 @@ type refImpl interface {
...
-//		assert.True(matches[0].Get(aIsAmbiguous).Equals(aIsAmbiguous.Field(0))
-//		assert.True(matches[1].Get(aIsAmbiguous).Equals(aIsAmbiguous.Field(1))
+//	aIsAmbiguous := arrow.NewSchema([]arrow.Field{
+//		{Name: "a", Type: arrow.PrimitiveTypes.Int32},
+//		{Name: "a", Type: arrow.PrimitiveTypes.Float32},
+//	})
+//	matches := FieldRefName("a").FindAll(aIsAmbiguous)
+//	assert.Len(matches, 2)
+//	assert.True(matches[0].Get(aIsAmbiguous).Equals(aIsAmbiguous.Field(0))
+//	assert.True(matches[1].Get(aIsAmbiguous).Equals(aIsAmbiguous.Field(1))
```
Comment (felipecrv) [inline go/arrow/compute/fieldref.go:309]: If this is markdown, code should be indented 4 spaces.

File: go/arrow/compute/fieldref.go [hunk 10]
```diff
@@ -282,31 +282,31 @@ type refImpl interface {
...
 // Nested fields can be referenced as well, given the schema:
 //
-// 		arrow.NewSchema([]arrow.Field{
-//			{Name: "a", Type: arrow.StructOf(arrow.Field{Name: "n", Type: arrow.Null})},
-//  		{Name: "b", Type: arrow.PrimitiveTypes.Int32},
-// 		})
+//			arrow.NewSchema([]arrow.Field{
+//				{Name: "a", Type: arrow.StructOf(arrow.Field{Name: "n", Type: arrow.Null})},
+//	 		{Name: "b", Type: arrow.PrimitiveTypes.Int32},
+//			})
```
Comment (felipecrv) [inline go/arrow/compute/fieldref.go:288]: Why the indentation increased here?

File: go/arrow/compute/fieldref.go [hunk 11]
```diff
@@ -346,17 +346,18 @@ func FieldRefList(elems ...interface{}) FieldRef {
...
-//		".beta[3]" => FieldRefList("beta", 3)
-//		"[5].gamma.delta[7]" => FieldRefList(5, "gamma", "delta", 7)
-//		".hello world" => FieldRefName("hello world")
-//		`.\[y\]\\tho\.\` => FieldRef(`[y]\tho.\`)
+//	".alpha" => FieldRefName("alpha")
+//	"[2]" => FieldRefIndex(2)
+//	".beta[3]" => FieldRefList("beta", 3)
+//	"[5].gamma.delta[7]" => FieldRefList(5, "gamma", "delta", 7)
+//	".hello world" => FieldRefName("hello world")
+//	`.\[y\]\\tho\.\` => FieldRef(`[y]\tho.\`)
```
Comment (felipecrv) [inline go/arrow/compute/fieldref.go:360]: Code should be indented or start end and with ` ``` `

File: go/arrow/internal/flatbuf/Binary.go [hunk 12]
```diff
@@ -22,7 +22,7 @@ import (
 	flatbuffers "github.com/google/flatbuffers/go"
 )
 
-/// Opaque binary data
+// / Opaque binary data
```
Comment (felipecrv) [inline go/arrow/internal/flatbuf/Binary.go:25]: Auto formatter noise? How can we prevent this?

File: go/arrow/internal/debug/log_on.go [hunk 13]
```diff
@@ -14,6 +14,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//go:build debug
```
Comment (felipecrv) [inline go/arrow/internal/debug/log_on.go:17]: no space after `//`?

File: go/arrow/internal/flatbuf/DictionaryBatch.go [hunk 14]
```diff
@@ -73,9 +73,9 @@ func (rcv *DictionaryBatch) Data(obj *RecordBatch) *RecordBatch {
 	return nil
 }
 
-/// If isDelta is true the values in the dictionary are to be appended to a
-/// dictionary with the indicated id. If isDelta is false this dictionary
-/// should replace the existing dictionary.
+// / If isDelta is true the values in the dictionary are to be appended to a
+// / dictionary with the indicated id. If isDelta is false this dictionary
+// / should replace the existing dictionary.
```
Comment (felipecrv) [inline go/arrow/internal/flatbuf/DictionaryBatch.go:78]: I think you should undo these changes `git checkout -p origin/main` and have a separate PR where we make the who Go codebase auto-formattable.

**Review by felipecrv — CHANGES_REQUESTED** [review]
There is a lot of noise in this PR. Can you split the refactoring you did from the cABI changes?

File: go/arrow/cdata/cdata_exports.go (hunk 7 above)
Comment (bkietz) [reply go/arrow/cdata/cdata_exports.go:345]: It is since there's a named return value

File: go/arrow/cdata/cdata.go (hunk 6 above)
Comment (bkietz) [reply go/arrow/cdata/cdata.go:765]: Nope, will replace with importFixedSizeBuffer

File: go/arrow/array/list.go (hunk 4 above)
Comment (bkietz) [reply go/arrow/array/list.go:1418]: RangeOfUsedValues was causing an error in IPC writing where it was being used to determine how much of the values array to write. This could drop values referenced by views under null bits, which later produced negative offsets. I was trying to rewrite this function to also consider views under null bits (less efficient for concat but still usable for IPC writing). I have reverted the change to RangeOfUsedValues and IPC writing no longer tries to slice the values array. This is less optimal for IPC writing but we can add optimizations back later, for now   I'd like to get the integration test passing.

File: go/arrow/internal/debug/log_on.go (hunk 13 above)
Comment (zeroshade) [reply go/arrow/internal/debug/log_on.go:17]: no, compiler directives like build constraints explicitly have no space after the `//`

File: go/arrow/internal/flatbuf/Binary.go (hunk 12 above)
Comment (bkietz) [reply go/arrow/internal/flatbuf/Binary.go:25]: I'll revert most of these

File: go/arrow/cdata/cdata.go [hunk 15]
```diff
@@ -654,6 +680,40 @@ func (imp *cimporter) importStringLike(offsetByteWidth int64) (err error) {
...
+
+	buffers := make([]*memory.Buffer, len(imp.cbuffers)-1)
+	// XXX couldn't figure out how to extract file_reader.go::releaseBuffers as a utility
+	defer func() {
+		for _, buf := range buffers {
+			if buf != nil {
+				buf.Release()
+			}
+		}
+	}()
```
Comment (zeroshade) [inline go/arrow/cdata/cdata.go:696]: we could export it (just make the first letter capitalized) but if we do that, it might make more sense for it to exist inside of `memory` as a utility rather than from the `ipc` package

File: go/arrow/internal/arrjson/arrjson.go [hunk 16]
```diff
@@ -1090,7 +1091,7 @@ func arrayFromJSON(mem memory.Allocator, dt arrow.DataType, arr Array) arrow.Arr
 	case arrow.BinaryViewDataType:
 		valids := validsToBitmap(validsFromJSON(arr.Valids), mem)
 		nulls := arr.Count - bitutil.CountSetBits(valids.Bytes(), 0, arr.Count)
-		headers := stringHeadersFromJSON(mem, !dt.IsUtf8(), arr.Data)
+		headers := stringHeadersFromJSON(mem, !dt.IsUtf8(), arr.Views)
```
Comment (zeroshade) [inline go/arrow/internal/arrjson/arrjson.go:1094]: was this a change in the integration testing spec?

File: go/arrow/internal/utils.go [hunk 17]
```diff
@@ -45,3 +45,17 @@ func HasValidityBitmap(id arrow.Type, version flatbuf.MetadataVersion) bool {
...
+
+// HasBufferSizesBuffer returns whether a given type has an extra buffer
+// in the C ABI to store the sizes of other buffers. Currently this is only
+// StringView and BinaryView.
+func HasBufferSizesBuffer(id arrow.Type) bool {
+	switch id {
+	case arrow.STRING_VIEW:
+		return true
+	case arrow.BINARY_VIEW:
+		return true
```
Comment (zeroshade) [inline go/arrow/internal/utils.go:57]:
```go
switch id {
case arrow.STRING_VIEW, arrow.BINARY_VIEW:
        return true
```

File: go/arrow/type_traits.go [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+		DayTimeInterval | MonthDayNanoInterval
+}
+
+type TemporalType interface {
+	Date32 | Date64 | Time32 | Time64 |
+		Timestamp | Duration | DayTimeInterval |
+		MonthInterval | MonthDayNanoInterval
+}
+
+func sliceAs[Out, T interface{}](b []T) []Out {
```
Comment (zeroshade) [inline go/arrow/type_traits.go:81]: just to be cleaner, use `any` instead of `interface{}`. I prefer it aesthetically

File: go/arrow/type_traits.go [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+  len_bytes := len(b) * int(unsafe.Sizeof(b[0]))
+  cap_bytes := cap(b) * int(unsafe.Sizeof(b[0]))
+
+  var z Out
+  len_out := len_bytes / int(unsafe.Sizeof(z))
+  cap_out := cap_bytes / int(unsafe.Sizeof(z))
```
Comment (zeroshade) [inline go/arrow/type_traits.go:87]: go typically prefers pascalCase rather than snake_case

File: go/arrow/type_traits.go [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+
+type TemporalType interface {
+	Date32 | Date64 | Time32 | Time64 |
+		Timestamp | Duration | DayTimeInterval |
+		MonthInterval | MonthDayNanoInterval
+}
+
+func sliceAs[Out, T interface{}](b []T) []Out {
+  len_bytes := len(b) * int(unsafe.Sizeof(b[0]))
+  cap_bytes := cap(b) * int(unsafe.Sizeof(b[0]))
```
Comment (zeroshade) [inline go/arrow/type_traits.go:83]: this'll blow up if `len(b) == 0`

File: go/arrow/type_traits.go [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+
+  var z Out
+  len_out := len_bytes / int(unsafe.Sizeof(z))
+  cap_out := cap_bytes / int(unsafe.Sizeof(z))
+
+	h := (*reflect.SliceHeader)(unsafe.Pointer(&b))
+  return unsafe.Slice((*Out)(unsafe.Pointer(h.Data)), cap_out)[:len_out]
```
Comment (zeroshade) [inline go/arrow/type_traits.go:90]: fix the indentation, and let's make this more portable by just doing `unsafe.Pointer(&b[0])` instead of going to the `reflect.SliceHeader`

File: go/arrow/type_traits.go [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+func GetValues[T FixedWidthType](data ArrayData, i int) []T {
```
Comment (zeroshade) [inline go/arrow/type_traits.go:93]: since we're exposing this now, we should add a proper comment docstring to it

File: go/arrow/type_traits.go [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+	if data.Buffers()[i] == nil || data.Buffers()[i].Len() == 0 {
+		return nil
+	}
+	return sliceAs[T](data.Buffers()[i].Bytes())[data.Offset():data.Offset()+data.Len()]
+}
+
+func GetOffsets[T int32 | int64](data ArrayData, i int) []T {
```
Comment (zeroshade) [inline go/arrow/type_traits.go:100]: same here and the ones below it

**Review by zeroshade — CHANGES_REQUESTED** [review]
in general this LGTM, just some nit picks

File: go/arrow/internal/arrjson/arrjson.go (hunk 16 above)
Comment (bkietz) [reply go/arrow/internal/arrjson/arrjson.go:1094]: Nope, will add

File: go/arrow/cdata/cdata.go (hunk 15 above)
Comment (bkietz) [reply go/arrow/cdata/cdata.go:696]: I'll try that

File: go/arrow/internal/arrjson/arrjson.go (hunk 16 above)
Comment (bkietz) [reply go/arrow/internal/arrjson/arrjson.go:1094]: https://github.com/apache/arrow/pull/39019/commits/b9237669d321e38bdeeffb1e675be5399a9575e8

File: dev/archery/archery/integration/datagen.py (hunk 3 above)
Comment (bkietz) [reply dev/archery/archery/integration/datagen.py]: https://github.com/apache/arrow/pull/39019/commits/03a12f9cf50dbb92184afd815a73454b138ce713

File: go/arrow/type_traits.go (hunk 21 above)
Comment (bkietz) [reply go/arrow/type_traits.go]: https://github.com/apache/arrow/pull/39019/commits/cb6a4efdd74c4b496572f2223be39f286bdcb5d1

File: go/arrow/array/list.go (hunk 4 above)
Comment (felipecrv) [reply go/arrow/array/list.go:1418]: This is why it's important to keep the documentation about the pre-condition `... && input.NullN() != input.Len()`. The concept of "minumum offset" is undefined when there are 0 offsets to look at (an empty set can't have a minimum) -- the caller can betetr decide what's appropriate to do when all values are null.

**Review by zeroshade — APPROVED** [review]

File: go/arrow/array/list.go [hunk 24]
```diff
@@ -1411,118 +1410,19 @@ func (b *baseListViewBuilder) UnmarshalJSON(data []byte) error {
...
+// Find the minimum offset+size in a LIST_VIEW/LARGE_LIST_VIEW array.
 //
 // Pre-conditions:
 //
-//	input.DataType() is ListViewType
-//	input.Len() > 0 && input.NullN() != input.Len()
-func maxListViewOffset32(input arrow.ArrayData) int {
-	inputOffset := input.Offset()
+//	input.DataType() is ListViewType if Offset=int32 or LargeListViewType if Offset=int64
+//	input.Len() > 0
```
Comment (felipecrv) [inline go/arrow/array/list.go:1418]:
```suggestion
//	input.Len() > 0 && input.NullN() != input.Len()
```

**Review by felipecrv — APPROVED** [review]

**Comment by zeroshade** [comment]
@pitrou any further issues here / remaining things to be addressed?

**Comment by zeroshade** [comment]
if there's no objections or no other changes requested I'll merge this at the end of the day
